specs/guaranteed-gas-market.md

```diff
@@ -0,0 +1,143 @@
+gas_target = GUARANTEED_GAS_LIMIT // ELASTICITY_MULTIPLIER
+
+# // implies floor division, however because gas_delta is always positive, it is the same as truncating (aka round to 0) division
+# If first deposit of this block, calculate the new basefee and store other info as well.
+if curr_num != block.number {
+    if curr_bought_gas == gas_target {
+        new_basefee := curr_basefee
+    } else if curr_bought_gas > gas_target {
+        gas_delta     := curr_bought_gas - gas_target
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        basefee_delta := max(basefee_delta, 1) # TODO: Why does 1559 have this asymmetry?
+        new_basefee   := curr_basefee + basefee_delta
+    } else {
+        gas_delta     := gas_target - curr_bought_gas
+        basefee_delta := gas_delta * curr_basefee // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+    # Fun fact, geth doesn't let the new_basefee get below 0 and while not in the EIP spec, we should add this as well.
+        new_basefee   := curr_basefee - basefee_delta
+    }
+    curr_basefee := new_basefee
+    curr_number := block.number
+    curr_bought_gas := 0
+   
+}
+
+curr_bought_gas += required_gas
+require(curr_bought_gas <= min(GUARANTEED_GAS_LIMIT, SANITY_GAS_LIMIT)
+gas_cost = requested_gas * curr_basefee
+
+burn(gas_cost) OR pay_to_contract(gas_cost) # Depends if payable or non-payable version
+
+pack_and_store(curr_basefee, curr_number, curr_bought_gas)
+```
+
+```python
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+
+prev_basefee, prev_num, prev_bought_gas = load_and_unpack_storage()
+
+gas_target_limit, gas_sanity_limit = load_and_unpack_storage2()
+
+
+gas_cost = requested_gas * prev_basefee
+
+now_basefee = prev_basefee
+now_num = block.number
+now_bought_gas = prev_bought_gas + requested_gas
+# update only if we are in a new block
+if now_num != prev_num:
+  now_bought_gas = requested_gas
+  # update basefee if we are using more or less than the target
+  if prev_bought_gas < gas_target:
+      gas_used_delta = prev_bought_gas - gas_target
+      base_fee_per_gas_delta = max(prev_basefee * gas_used_delta // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR, 1)
+      now_basefee = prev_basefee + base_fee_per_gas_delta
+  elif prev_bought_gas > gas_target:
+      gas_used_delta = gas_target - prev_bought_gas
+      base_fee_per_gas_delta = prev_basefee * gas_used_delta // gas_target // BASE_FEE_MAX_CHANGE_DENOMINATOR
+      now_basefee = prev_basefee - base_fee_per_gas_delta
+
+  # TODO: we can reduce now_basefee if now_num > prev_num by multiple blocks,
+  #   since that means we had some rest on L2
+  
+  # optional: maybe reduce gas cost a little bit for spending gas on above math work
+
+require(now_bought_gas < gas_sanity_limit)  # limit how much can be bought per L1 block
+
+required_lockup = mint + (requested_gas * now_basefee)
+require(msg.value >= required_lockup)
+
+pack_and_store(now_basefee, now_num, now_bought_gas)
+```
```

Comment:
Didn't re-review this, but is it normal there are two copies of the pseudocode? 

Same comment as before, not really sure what `pack_and_store` is.

Reply:
We'll decide on one copy before merging. `pack_and_store` is about storing multiple values in a single storage slot (as part of solidity).

Reply:
What's the difference between both copies?

Reply:
There are some slight differences in the logic, need to reconcile which is the desired behavior. I have a preference for the python, as I find it easier to read